### PR TITLE
Prevent adding duplicate props, fix duplicate props on startup (if broken)

### PR DESCRIPTION
**Pull request by etiennedi**
### What's being changed:
* prevent adding duplicate props through better validation
* add missing tests for ⬆️ 
* if there are duplicate props at startup, de-duplicate them before trying to init the same prop twice

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/schema/remove_duplicate_props.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	return nil
+}
+
+func hasDuplicateProps(props []*models.Property) bool {
+	counts := map[string]int{}
+
+	for _, prop := range props {
+		counts[prop.Name]++
```
Comment (aliszka) [inline usecases/schema/remove_duplicate_props.go:35]: should propname here be case sensitive or insensitive?

would be simpler to check key presence in the map instead counting all occurrences

**Review by aliszka — APPROVED** [review]

File: usecases/schema/remove_duplicate_props.go (hunk 1 above)
Comment (etiennedi) [reply usecases/schema/remove_duplicate_props.go]: both good points, addressed in the latest commit.

File: usecases/schema/manager.go [hunk 2]
```diff
@@ -206,6 +206,15 @@ func (m *Manager) loadOrInitializeSchema(ctx context.Context) error {
 		return fmt.Errorf("migrate schema: %w", err)
 	}
 
+	// There was a bug that allowed adding the same prop multiple times. This
+	// leads to a race at startup. If an instance is already affected by this,
+	// this step can remove the duplicate ones.
+	//
+	// See https://github.com/weaviate/weaviate/issues/2609
+	if err := m.removeDuplicatePropsIfPresent(ctx); err != nil {
```
Comment (redouan-rhazouani) [inline usecases/schema/manager.go:214]: It appears `m.removeDuplicatePropsIfPresent()` doesn't need a context and it returns nil always

File: usecases/schema/remove_duplicate_props.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/entities/models"
+)
+
+func (m *Manager) removeDuplicatePropsIfPresent(ctx context.Context) error {
+	for _, c := range m.state.ObjectSchema.Classes {
+		if hasDuplicateProps(c.Properties) {
```
Comment (redouan-rhazouani) [inline usecases/schema/remove_duplicate_props.go:24]: We can save the call to `hasDuplicateProps()` and instead directly call m.deduplicateProps() (see comment below about in place version)

File: usecases/schema/remove_duplicate_props.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	for _, count := range counts {
+		if count > 1 {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (m *Manager) deduplicateProps(orig []*models.Property,
```
Comment (redouan-rhazouani) [inline usecases/schema/remove_duplicate_props.go:46]: This could be done in-place

**Review by redouan-rhazouani — APPROVED** [review]

**Pull request merged**